File: DiagramsDesktop/Math/ViewportMath.cs
using System;
using DiagramsDesktop.Models;

namespace DiagramsDesktop.Math;

public static class ViewportMath
{
    public static (double dX, double dY) CalculatePan(double deltaX, double deltaY, double zoom)
    {
        if (zoom <= 0) return (0, 0);
        return (
            dX: -(deltaX / zoom),
            dY: deltaY / zoom // Y-flip: screen Y moves down, world Y must increase (or vice versa depending on logic)
        );
    }

    public static (double zoom, double? newVCX, double? newVCY) CalculateZoom(
        double currentZoom,
        double direction,
        double anchorScreenX,
        double anchorScreenY,
        double canvasWidth,
        double canvasHeight,
        CanvasModel canvas)
    {
        double step = CanvasConfig.ZoomStep;
        double newZoom = direction > 0
            ? System.Math.Min(CanvasConfig.MaxZoom, currentZoom + step)
            : System.Math.Max(CanvasConfig.MinZoom, currentZoom - step);

        if (System.Math.Abs(newZoom - currentZoom) < 1e-9)
        {
            return (newZoom, null, null);
        }

        double sCenterX = canvasWidth / 2.0;
        double sCenterY = canvasHeight / 2.0;

        double dx = anchorScreenX - sCenterX;
        double dy = anchorScreenY - sCenterY;

        double newVCX = canvas.ViewportCenterX + dx * (1.0 / currentZoom - 1.0 / newZoom);
        double newVCY = canvas.ViewportCenterY - dy * (1.0 / currentZoom - 1.0 / newZoom);

        return (newZoom, newVCX, newVCY);
    }

    public static (double x, double y) WorldToScreen(
        double worldX,
        double worldY,
        CanvasModel canvas,
        double canvasWidth,
        double canvasHeight)
    {
        if (canvas == null) return (0, 0);

        double zoom = canvas.ZoomScale;
        double vCenterX = canvas.ViewportCenterX;
        double vCenterY = canvas.ViewportCenterY;

        double sCenterX = canvasWidth / 2.0;
        double sCenterY = canvasHeight / 2.0;

        double screenX = sCenterX + (worldX - vCenterX) * zoom;
        double screenY = sCenterY - (worldY - vCenterY) * zoom; // Y-flip

        return (screenX, screenY);
    }

    public static (double x, double y) ScreenToWorld(
        double screenX,
        double screenY,
        CanvasModel canvas,
        double canvasWidth,
        double canvasHeight)
    {
        if (canvas == null) return (0, 0);

        double zoom = canvas.ZoomScale;
        double vCenterX = canvas.ViewportCenterX;
        double vCenterY = canvas.ViewportCenterY;

        double sCenterX = canvasWidth / 2.0;
        double sCenterY = canvasHeight / 2.0;

        double worldX = (screenX - sCenterX) / zoom + vCenterX;
        double worldY = -(screenY - sCenterY) / zoom + vCenterY; // Y-flip

        return (worldX, worldY);
    }
}

File: publish/DiagramsDesktop/wwwroot/js/viewport-math.js
/**
 * viewport-math.js
 * =================
 * Handles logic for panning and zooming calculations.
 */

'use strict';

const ViewportMath = (() => {

  function calculatePan(deltaX, deltaY, zoom) {
    // Panning moves the viewport in the direction of the drag.
    // Screen X moves right → world X moves right → viewport center increases → objects move left: correct.
    // Screen Y moves down  → world Y is UpPositive, so viewport center must DECREASE → objects appear to move up: correct.
    // BUG-08 fix: dY must be negated to respect the Y-flip between screen and world space.
    return {
      dX: -(deltaX / zoom),
      dY:  (deltaY / zoom)   // BUG-08 fixed: was incorrect sign; both axes now pan correctly
    };
  }

  function calculateZoom(currentZoom, direction, anchorScreenX, anchorScreenY, canvasWidth, canvasHeight) {
    const step = CanvasConfig.ZOOM_STEP;
    const newZoom = direction > 0 
      ? Math.min(CanvasConfig.MAX_ZOOM, currentZoom + step)
      : Math.max(CanvasConfig.MIN_ZOOM, currentZoom - step);
    
    // Zoom toward anchor point (mouse position)
    // To keep world point under mouse fixed:
    // W_old = (S - SC)/Zoom_old + VC_old
    // W_new = (S - SC)/Zoom_new + VC_new
    // Setting W_old = W_new and solving for VC_new:
    // VC_new = VC_old + (S - SC) * (1/Zoom_old - 1/Zoom_new)
    
    if (newZoom === currentZoom) return { zoom: newZoom };

    const canvas = CanvasState.getCanvas();
    const sCenterX = canvasWidth / 2;
    const sCenterY = canvasHeight / 2;

    const dx = anchorScreenX - sCenterX;
    const dy = anchorScreenY - sCenterY;

    const newVCX = canvas.ViewportCenterX + dx * (1/currentZoom - 1/newZoom);
    // Y-flip handled by using dy correctly
    const newVCY = canvas.ViewportCenterY - dy * (1/currentZoom - 1/newZoom);

    return {
      zoom: newZoom,
      newVCX,
      newVCY
    };
  }

  return { calculatePan, calculateZoom };

})();
